private/bufpkg/bufcheck/buflint/internal/buflintvalidate/util.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package buflintvalidate
+
+import (
+	"reflect"
+)
+
+// validateNumberField asserts that the given numbers are valid.
+func validateNumberField[T any](m *validateField, in, notIn int, ci, lti, ltei, gti, gtei T) {
+	m.checkIns(in, notIn)
+
+	c := reflect.ValueOf(ci)
+	lt, lte := reflect.ValueOf(lti), reflect.ValueOf(ltei)
+	gt, gte := reflect.ValueOf(gti), reflect.ValueOf(gtei)
+
+	m.assertf(c.IsNil() ||
+		in == 0 && notIn == 0 &&
+			lt.IsNil() && lte.IsNil() &&
+			gt.IsNil() && gte.IsNil(),
+		"`const` can be the only rule on a field",
+	)
+
+	m.assertf(in == 0 ||
+		lt.IsNil() && lte.IsNil() &&
+			gt.IsNil() && gte.IsNil(),
+		"cannot have both `in` and range constraint rules on the same field",
+	)
+
+	if !lt.IsNil() {
+		m.assertf(gt.IsNil() || !reflect.DeepEqual(lti, gti),
+			"cannot have equal `gt` and `lt` rules on the same field")
```

Comment:
We haven't used backticks in any of our other lint rules - generally, if we need to quote something, we use double-quotes. I won't comment about this in rest of PR, but obviously take care of it globally.

Reply:
resolved in [a3c77f8](https://github.com/bufbuild/buf/pull/2377/commits/a3c77f8869438dc13611808c7e3d10bcdbf8dc1f) and [182d2e2](https://github.com/bufbuild/buf/pull/2377/commits/182d2e21541d6084e75fd345150fa990cde60d62)